### DISCUSSION TITLE
DOMContentLoaded Event listener

### DISCUSSION THREAD

**Discussion by Tsenaye** [vercel/next.js#36622] [Help] [answered]
I'm trying to add a DOMContentLoaded event listener on one of my pages, however it seems this is not possible with nextJS. The document is always in readyState 'complete' when the page renders. I've tried to grab hold of these event in the custom _document as well as the custom _app components to no avail. Please assist if you can.

**Comment by icyJoseph** [comment, marked as answer]
This is tricky, because this event triggers very early...

You could do:

```js
 <script
                dangerouslySetInnerHTML={{
                    __html: `
            window.addEventListener('DOMContentLoaded', () => {
                console.log('DOM fully loaded and parsed');
            });
    `,
                }}
            ></script>
```

In the document Head for example, but you'd lose JavaScript context, because the __html prop is a string.

**Reply by Tsenaye** [reply]
Thanks for the response. I'm sorry, I didn't provide enough context. The event listener is meant to run a function that renders a secure payment button on the page. This function is loaded from an external script (from our payment gateway providers). Is it safe to have something like this in dangerouslySetInnerHTML? Btw, I finally managed to render the button on the page after days of work, so thank you very much on that score, sir.

**Reply by icyJoseph** [reply]
Yeah, I mean, it's not optimal, but it works, https://nextjs.org/docs/basic-features/script#inline-scripts. Long story short, it is not possible (to my knowledge) to inject code into template literals (``), they are not like eval. If the string came in as a prop, then we'd be really exposing ourselves.

I'd recommend exploring https://nextjs.org/docs/basic-features/script#afterinteractive, using next/script

```jsx
<Script
  strategy="afterInteractive"
  dangerouslySetInnerHTML={{
    __html: `
    (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
    new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
    j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
    'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
    })(window,document,'script','dataLayer', 'GTM-XXXXXX');
  `,
  }}
/>
```

And also load the external script via next/script, perhaps something like this, https://nextjs.org/docs/basic-features/script#executing-code-after-loading-onload:

```jsx
import { useState } from 'react'
import Script from 'next/script'

export default function Home() {
  const [stripe, setStripe] = useState(null)

  return (
    <>
      <Script
        id="stripe-js"
        src="https://js.stripe.com/v3/"
        onLoad={() => {
          setStripe({ stripe: window.Stripe('pk_test_12345') })
         // or something else
        }}
      />
    </>
  )
}
```

After all, what matters is that the script is ready, and then run the logic. My reasoning here is that the DOMContentLoaded event fires way to early, and simply invoking the external script via next/script (perhaps with the beforeInteractive option), should be fast enough for human eyes.

And don't forget to mark as resolved :) - Glad to help out!

**Reply by Tsenaye** [reply]
Thank you sir, the second option is what worked. But this only worked when I added the script to the document head. Like so
`<script src="https://js.stripe.com/v3/"></script>`

**Reply by icyJoseph** [reply]
I was reviewing this, and noticed that this should also work, but when testing it earlier on I had made a typo 😊 :

```tsx
_app.jsx
import type { AppProps } from "next/app";

if (typeof window !== "undefined") {
  // window.addEventListener("load", () => console.log("LOADED"));
  document.addEventListener("DOMContentLoaded", () =>
    console.log("DOMContentLoaded")
  );
}

function MyApp({ Component, pageProps }: AppProps) {
  return <Component {...pageProps} />;
}

export default MyApp;
```

Just place that `if` statement in the `_app.tsx` module. 

If you have time, could you confirm? It works on my issue testing repository.

**Reply by ljubar** [reply]
works!
